File: bindings/lua/Polycode/Texture.lua
require "Polycode/Resource"

class "Texture" (Resource)

Texture.FILTERING_NEAREST = 0
Texture.FILTERING_LINEAR = 1

function Texture:__getvar(name)
	if name == "clamp" then
		return Polycode.Texture_get_clamp(self.__ptr)
	elseif name == "textureData" then
		local retVal = Polycode.Texture_get_textureData(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["char"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "type" then
		return Polycode.Texture_get_type(self.__ptr)
	elseif name == "filteringMode" then
		return Polycode.Texture_get_filteringMode(self.__ptr)
	elseif name == "createMipmaps" then
		return Polycode.Texture_get_createMipmaps(self.__ptr)
	elseif name == "anisotropy" then
		return Polycode.Texture_get_anisotropy(self.__ptr)
	elseif name == "framebufferTexture" then
		return Polycode.Texture_get_framebufferTexture(self.__ptr)
	elseif name == "depthTexture" then
		return Polycode.Texture_get_depthTexture(self.__ptr)
	end
	if Resource["__getvar"] ~= nil then
		return Resource.__getvar(self, name)
	end
end

function Texture:__setvar(name,value)
	if name == "clamp" then
		Polycode.Texture_set_clamp(self.__ptr, value)
		return true
	elseif name == "textureData" then
		Polycode.Texture_set_textureData(self.__ptr, value.__ptr)
		return true
	elseif name == "type" then
		Polycode.Texture_set_type(self.__ptr, value)
		return true
	elseif name == "filteringMode" then
		Polycode.Texture_set_filteringMode(self.__ptr, value)
		return true
	elseif name == "createMipmaps" then
		Polycode.Texture_set_createMipmaps(self.__ptr, value)
		return true
	elseif name == "anisotropy" then
		Polycode.Texture_set_anisotropy(self.__ptr, value)
		return true
	elseif name == "framebufferTexture" then
		Polycode.Texture_set_framebufferTexture(self.__ptr, value)
		return true
	elseif name == "depthTexture" then
		Polycode.Texture_set_depthTexture(self.__ptr, value)
		return true
	end
	if Resource["__setvar"] ~= nil then
		return Resource.__setvar(self, name, value)
	else
		return false
	end
end
function Texture:reloadResource()
	local retVal =  Polycode.Texture_reloadResource(self.__ptr)
end

function Texture:setImageData(data)
	local retVal = Polycode.Texture_setImageData(self.__ptr, data.__ptr)
end

function Texture:getTextureData()
	local retVal =  Polycode.Texture_getTextureData(self.__ptr)
	if retVal == nil then return nil end
	local __c = _G["char"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Texture:getWidth()
	local retVal =  Polycode.Texture_getWidth(self.__ptr)
	return retVal
end

function Texture:getHeight()
	local retVal =  Polycode.Texture_getHeight(self.__ptr)
	return retVal
end

function Texture:setCreateMipmaps(createMipmapsIn)
	local retVal = Polycode.Texture_setCreateMipmaps(self.__ptr, createMipmapsIn)
end

function Texture:getCreateMipmaps()
	local retVal =  Polycode.Texture_getCreateMipmaps(self.__ptr)
	return retVal
end

function Texture:__delete()
	if self then Polycode.delete_Texture(self.__ptr) end
end

File: bindings/javascript/Polycode/Texture.js
require('Polycode/Resource')

function Texture() {
	Object.defineProperties(this, {
		'clamp': { enumerable: true, configurable: true, get: Texture.prototype.__get_clamp, set: Texture.prototype.__set_clamp},
		'textureData': { enumerable: true, configurable: true, get: Texture.prototype.__get_textureData, set: Texture.prototype.__set_textureData},
		'type': { enumerable: true, configurable: true, get: Texture.prototype.__get_type, set: Texture.prototype.__set_type},
		'filteringMode': { enumerable: true, configurable: true, get: Texture.prototype.__get_filteringMode, set: Texture.prototype.__set_filteringMode},
		'createMipmaps': { enumerable: true, configurable: true, get: Texture.prototype.__get_createMipmaps, set: Texture.prototype.__set_createMipmaps},
		'anisotropy': { enumerable: true, configurable: true, get: Texture.prototype.__get_anisotropy, set: Texture.prototype.__set_anisotropy},
		'framebufferTexture': { enumerable: true, configurable: true, get: Texture.prototype.__get_framebufferTexture, set: Texture.prototype.__set_framebufferTexture},
		'depthTexture': { enumerable: true, configurable: true, get: Texture.prototype.__get_depthTexture, set: Texture.prototype.__set_depthTexture}
	})
}

Texture.FILTERING_NEAREST = 0
Texture.FILTERING_LINEAR = 1

Texture.prototype = Object.create(Resource.prototype)

Texture.prototype.__get_clamp = function() {
	return Polycode.Texture__get_clamp(this.__ptr)
}

Texture.prototype.__set_clamp = function(val) {
	Polycode.Texture__set_clamp(this.__ptr, val)
}

Texture.prototype.__get_textureData = function() {
	var retVal = new char()
	retVal.__ptr = 	Polycode.Texture__get_textureData(this.__ptr)
	return retVal
}

Texture.prototype.__set_textureData = function(val) {
	Polycode.Texture__set_textureData(this.__ptr, val.__ptr)
}

Texture.prototype.__get_type = function() {
	return Polycode.Texture__get_type(this.__ptr)
}

Texture.prototype.__set_type = function(val) {
	Polycode.Texture__set_type(this.__ptr, val)
}

Texture.prototype.__get_filteringMode = function() {
	return Polycode.Texture__get_filteringMode(this.__ptr)
}

Texture.prototype.__set_filteringMode = function(val) {
	Polycode.Texture__set_filteringMode(this.__ptr, val)
}

Texture.prototype.__get_createMipmaps = function() {
	return Polycode.Texture__get_createMipmaps(this.__ptr)
}

Texture.prototype.__set_createMipmaps = function(val) {
	Polycode.Texture__set_createMipmaps(this.__ptr, val)
}

Texture.prototype.__get_anisotropy = function() {
	return Polycode.Texture__get_anisotropy(this.__ptr)
}

Texture.prototype.__set_anisotropy = function(val) {
	Polycode.Texture__set_anisotropy(this.__ptr, val)
}

Texture.prototype.__get_framebufferTexture = function() {
	return Polycode.Texture__get_framebufferTexture(this.__ptr)
}

Texture.prototype.__set_framebufferTexture = function(val) {
	Polycode.Texture__set_framebufferTexture(this.__ptr, val)
}

Texture.prototype.__get_depthTexture = function() {
	return Polycode.Texture__get_depthTexture(this.__ptr)
}

Texture.prototype.__set_depthTexture = function(val) {
	Polycode.Texture__set_depthTexture(this.__ptr, val)
}


Texture.prototype.reloadResource = function() {
	Polycode.Texture_reloadResource(this.__ptr)
}

Texture.prototype.setImageData = function(data) {
	Polycode.Texture_setImageData(this.__ptr, data.__ptr)
}

Texture.prototype.getTextureData = function() {
	var retVal = new char()
	retVal.__ptr = Polycode.Texture_getTextureData(this.__ptr)
	return retVal
}

Texture.prototype.getWidth = function() {
	return Polycode.Texture_getWidth(this.__ptr)
}

Texture.prototype.getHeight = function() {
	return Polycode.Texture_getHeight(this.__ptr)
}

Texture.prototype.setCreateMipmaps = function(createMipmapsIn) {
	Polycode.Texture_setCreateMipmaps(this.__ptr, createMipmapsIn)
}

Texture.prototype.getCreateMipmaps = function() {
	return Polycode.Texture_getCreateMipmaps(this.__ptr)
}
